### PR TITLE
[FLINK-32298][network] Fix the bug that totalWrittenBytes of BufferWritingResultPartition misses some data

### PR DESCRIPTION
**Pull request by 1996fanrui**
## What is the purpose of the change

The outputQueueSize indicates `The real size of queued output buffers in bytes.`, so it shouldn't be negative. However, it may be negative in some cases.

You can get detailed background from FLINK-32298 .

## Brief change log

Fix the bug that totalWrittenBytes of BufferWritingResultPartition misses some data, including:

- PipelinedSubpartition#finishReadRecoveredState writes the `EndOfChannelStateEvent` even if the unaligned checkpoint is disable
- PipelinedSubpartition#addRecovered writes channel state(if the job recovered from unaligned checkpoint, the outputQueueSize is totally wrong)
- PipelinedSubpartition#finish writes the `EndOfPartitionEvent`

## Verifying this change

Improved the `ResultPartitionTest#testSizeOfQueuedBuffers`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`:no
  - The serializers:no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not documented

### CODE REVIEW

**Comment by 1996fanrui** [comment]
Hi @reswqa  @akalash , would you mind help review this PR in your free time? thanks~

**Review by akalash — COMMENTED** [review]
@1996fanrui , thanks for this fix. I  a little bit worry about the second commit it did quite of changes which I am not sure about. I mean it definitely wasn't a good implementation before but we need to think about if it makes sense to change it in such a way. I also understand that the main idea of this fix is to do all the calculations inside Partition which means we should do all work with subpartition inside Partition. That is good but maybe we can do it with fewer changes. 
I'm still in the process of the review so I will think about the current solution more.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/RecoveredChannelStateHandler.java
Comment (1996fanrui) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/RecoveredChannelStateHandler.java:214]: Do you mean we should call `getCheckpointedResultPartition` based on the `subpartitionInfo` instead of mappedSubpartition?

If yes, sounds make sense. And updated at the last commit.

**Review by 1996fanrui — COMMENTED** [review]
Hi @akalash , thanks for the review. I have updated all comments.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/RecoveredChannelStateHandler.java
Comment (akalash) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/RecoveredChannelStateHandler.java:209]: good. It's better now

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/BufferWritingResultPartition.java
Comment (akalash) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/BufferWritingResultPartition.java:68]: Ideally, `totalWrittenBytes` should be changed only in one class `BufferWritingResultPartition`. So maybe we should forbid `inherited classes` to write to `subpartitions` directly. But honestly, I don't know how to do it easily since the current implementation of `inherited classes` use `subpartitions` a lot and we can not change `subpartitions` visibility from `protected` to `private` . I mean we can create `addToSubpartition` method in `BufferWritingResultPartition` and use it everywhere(and maybe it will be better than now) but it still doesn't protect us from possible future bugs if somebody decides to use `subpartitions#add` directly in `inherited classes`.

Anyway, if you don't think that it will make things worse, maybe we indeed for now create the `add`(or `addToSubpartition`) method in `BufferWritingResultPartition` and will use it everywhere. At least it will reduce the number of places where we update `totalWrittenBytes`. Unfortunatelly, I didn't come up with any other solution until now.

**Review by akalash — COMMENTED** [review]
It's already pretty good right now. I am just trying to come up with how we can decrease the spreading of `totalWrittenBytes`. I left some comments about it and I am happy to hear your opinion about that. But anyway, the current solution is good enough so if we won't find anything better we can leave it as is.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/BufferWritingResultPartition.java
Comment (1996fanrui) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/BufferWritingResultPartition.java]: Hi @akalash , thanks a lot for your very patient review again.

>  I mean we can create addToSubpartition method in BufferWritingResultPartition and use it everywhere(and maybe it will be better than now)

I think it will be better than now as well, and this is the only better solution I can think of right now.

Updated by the last commit.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/RecoveredChannelStateHandler.java
Comment (akalash) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/RecoveredChannelStateHandler.java]: No, I don't have any good suggestions for now. I just wanted to emphasize that `ResultSubpartitionRecoveredStateHandler` uses `ResultPartitionWriter` for the writer while it actually needs `CheckpointedResultPartition`. As a result, it requires the explicit cast to this type. 
So it doesn't look clean solution but since it was here before and I don't actually think that we should try to improve it in this fix. We just can leave it as is.

**Review by akalash — APPROVED** [review]
@1996fanrui , thanks again for these changes and the waiting for my review. Now LGTM. Don't forget to squash your commits before merging as soon as the tests are green. It also makes sense to check the benchmarks since this code is sensitive to performance(at least you can take a look at benchmarks after the merge to be sure that nothing was degraded)

**Comment by 1996fanrui** [comment]
Thanks for your quick review again!

> Don't forget to squash your commits before merging as soon as the tests are green. 

Done.

> It also makes sense to check the benchmarks since this code is sensitive to performance(at least you can take a look at benchmarks after the merge to be sure that nothing was degraded.

Sorry, I don't know why this PR is sensitive to performance, I may have missed some background. As I understand, it just add buffer through ResultPartition instead of adding buffer to Subpartition directly.

Could you give more information? thanks a lot.

BTW, do you think this fix should be backported to 1.17 and 1.16? It's a bug, however it only affects the metric.

**Comment by 1996fanrui** [comment]
@akalash Thanks for your quick review again!

> Don't forget to squash your commits before merging as soon as the tests are green. 

Done.

> It also makes sense to check the benchmarks since this code is sensitive to performance(at least you can take a look at benchmarks after the merge to be sure that nothing was degraded.

Sorry, I don't know why this PR is sensitive to performance, I may have missed some background. As I understand, it just add buffer through ResultPartition instead of adding buffer to Subpartition directly.

Could you give more information? thanks a lot.

BTW, do you think this fix should be backported to 1.17 and 1.16? It's a bug, however it only affects the metric.

**Comment by akalash** [comment]
`BufferWritingResultPartition` is performance critical class since it is on the hot path. But your changes mostly touched not so 'hot' things like recovery and I don't think that it will be any degradation. It's why I don't think that you should do benchmarking for this PR but since `BufferWritingResultPartition` was touched anyhow I suggest taking a look at the benchmarks the next day after merging this PR to be 100% sure that it doesn't bring any degradation

> BTW, do you think this fix should be backported to 1.17 and 1.16?

Good question. Officially we should support the previous version but I'm not fully sure that it makes sense to do it for metrics improvement. So I think it won't be a mistake if you backport it but I think it is also ok if you merge only to master. (we can always backport later if somebody decided that it is worth it)

**Comment by 1996fanrui** [comment]
Thanks for the clarification, I will follow the benchmake tomorrow.

CI passed, merging~

**Comment by 1996fanrui** [comment]
Hi @akalash , I have followed the benchmark result, and I didn't found this pr effects them.

And I will follow it again next week.

**Pull request merged**
